### ISSUE TITLE
UI Git builds fail

### ISSUE THREAD

**Issue by abonillabeeche** [epinio/ui#148]
**Describe the bug**
App Deployment fails when done via the UI, but succeeds from the CLI. Perhaps the web UI is requesting a token that expires while the build is in progress.

**To Reproduce**

1. Install Epinio 1.0
2. Open the UI, select Create an App, use the Git option
3. https://github.com/abonillabeeche/dizzylizard for the URL and Master branch
4. Create the App, it will fail on build.

**Expected behavior**
Build should be Deployed - as the build succeeds.

**Logs**
When it succeeds 👍🏼

🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Paketo Procfile Buildpack 5.1.2 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack   https://github.com/paketo-buildpacks/procfile 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack   Process types: 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack     web: python -m SimpleHTTPServer $PORT 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack ===> EXPORTING 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding layer 'paketo-buildpacks/ca-certificates:helper' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding 1/1 app layer(s) 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding layer 'launcher' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding layer 'config' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding label 'io.buildpacks.lifecycle.metadata' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding label 'io.buildpacks.build.metadata' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding label 'io.buildpacks.project.metadata' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Saving registry.epinio.svc.cluster.local:5000/apps/workspace-dizzy:b725e78de1f74bf4... 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack *** Images (sha256:ffd4e53bbc5f8b578d4c41e67f9c61d06a936202270325de85d4e5c3482961c9): 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack       registry.epinio.svc.cluster.local:5000/apps/workspace-dizzy:b725e78de1f74bf4 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding cache layer 'paketo-buildpacks/node-engine:node' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding cache layer 'paketo-buildpacks/npm-install:build-modules' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding cache layer 'paketo-buildpacks/npm-install:npm-cache' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding cache layer 'paketo-buildpacks/node-module-bom:cyclonedx-node-module' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack _ _ __ ___ _____ Done 

Deploying application ...
..............................
🕞  Creating application resources
✔️  App is online.
Name: dizzy
Namespace: workspace
Builder Image: 
Routes: 
1: https://dizzy.dev.5glinux.com 

but when it fails, via the UI, the log ends at...

🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack Adding cache layer 'paketo-buildpacks/node-module-bom:cyclonedx-node-module' 
🕞  [stage-workspace-dizzy--1adf17e1992a52d9450aaaed90687ae538a2ddhb] buildpack _ _ __ ___ _____ Done 

To follow the server logs you can use:

`kubectl logs -n epinio -l app.kubernetes.io/name=epinio-server -c epinio-server -f`

2022/07/05 19:23:50 EpinioServer: "msg"="Error" "error"="token expired"
2022/07/05 19:24:19 EpinioServer: "msg"="ERROR" "error"="token expired" "requestId"="aa3695b5-0307-46ad-8c1d-f43a7901e3b6" "origin"="/wapi/v1/namespaces/workspace/staging/4768245a36aa69c4/logs?follow=true&authtoken=eyJhbGciOiJSUzM4NCIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJlcGluaW8tc2VydmVyIiwiZXhwIjoxNjU3MDQ4NTQ5LCJ1c2VyIjoiYWRtaW4ifQ.VouuOmqfTFz2v0mugjoXB6sJyBRzjc0QgTu_KFl6Ml61hPdSzJh11hZM21pBxKb6ITcXJExNXAy7zHItuMw4lfE38EwRjikSNeI8hUDPMpw-Qe5tcdpEGGBsro_IPKe_1lH2Fy-XH6zMBbmhcprX3jK266Qw1-TY7LwtVe2gQhr2cUxMqbfIzmCnSYojav-VCRXSyhU4vwB-aV8vJT655n6BQKaKIW6jdE_uQ_I62h-0Z2i4MqMFSP1DE1IPtz8J1bOKyuzYbvfDao0hm_wd9Rx9apbxgZHodAcR-H805nR-m_Udo1V1ARKiLtrGsXE49rE6Eeq9722F-3toEidlLg&sockId=29" "error"="token expired"
2022/07/05 19:24:19 EpinioServer: "msg"="Error" "error"="token expired"
2022/07/05 19:24:48 EpinioServer: "msg"="ERROR" "error"="token expired" "requestId"="61e063ef-a9e9-463e-981a-2493c5bd3c9c" "origin"="/wapi/v1/namespaces/workspace/staging/4768245a36aa69c4/logs?follow=true&authtoken=eyJhbGciOiJSUzM4NCIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJlcGluaW8tc2VydmVyIiwiZXhwIjoxNjU3MDQ4NTQ5LCJ1c2VyIjoiYWRtaW4ifQ.VouuOmqfTFz2v0mugjoXB6sJyBRzjc0QgTu_KFl6Ml61hPdSzJh11hZM21pBxKb6ITcXJExNXAy7zHItuMw4lfE38EwRjikSNeI8hUDPMpw-Qe5tcdpEGGBsro_IPKe_1lH2Fy-XH6zMBbmhcprX3jK266Qw1-TY7LwtVe2gQhr2cUxMqbfIzmCnSYojav-VCRXSyhU4vwB-aV8vJT655n6BQKaKIW6jdE_uQ_I62h-0Z2i4MqMFSP1DE1IPtz8J1bOKyuzYbvfDao0hm_wd9Rx9apbxgZHodAcR-H805nR-m_Udo1V1ARKiLtrGsXE49rE6Eeq9722F-3toEidlLg&sockId=30" "error"="token expired"
2022/07/05 19:24:48 EpinioServer: "msg"="Error" "error"="token expired"

Epinio 1.0.0 - the push fail  is only visible via the UI

Create Application | The Application will be created ready to deploy source to | Success
-- | -- | --
Fetch | Fetch the files at the git repository's branch | Success
Build | Build the source for the Application | Fail
[object Object]
Deploy | Deploy and start the Application | Pending

**Comment by richard-cox** [comment]
As discussed in planning. This may be an issue where a stale token is being used by the UI when opening a socket connection or an issue where the token supplied by Epinio is stale/incorrectly created.

I'll validate that there's nothing amiss UI side and if applicable transfer back for the Epinio team to look in to

**Comment by richard-cox** [comment]
From the investigation above, the token used to stream the staging logs over socket has expired (token expires `19:15`, error is later `19:24`).

Before opening any socket the UI will request a token. So either the socket streaming the staging logs is still open but token expires (going from discussions in planning i don't think this is possible) or the UI opens a socket with a stale token.

I tried to reproduce the latter a number of ways, and found only one where it could happen.

1) In UI bring up the logs for an application
2) In kube hosting epinio kill the epinio api pod
  a) The UI socket connect is broken
  b) Kube hosting epinio brings back up epinio api pod
  c) The UI tries to reconnect using the same URL

At this point, even if the token hasn't expired, I get a `token expired` error in the api logs

``` 
  2022/08/09 10:42:19 EpinioServer: "msg"="ERROR" "error"="token expired" "requestId"="883dc615-e07b-4300-b749-856da52025c0" "origin"="/wapi/v1/namespaces/richard/applications/dizzylizard/logs?follow=true&auth token=eyJhbGciOiJSUzM4NCIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJlcGluaW8tc2VydmVyIiwiZXhwIjoxNjYwMDQxNjcyLCJ1c2VyIjoiYWRtaW4ifQ.M90Q_XsYuebWwsoK0HOi6Ol1UmX4D1RqXCHvkI0AYaLIazCLukjoVP-olOxYLnXpaMZVVy2VQSThrselOcirZQOX9rKX4bP9sdZQJuSI-kk4lCxv725jV3_elBupuZQxyhv4I2tg5c_8WjYiAUeGlBZ-d4-d0RVS9QveRcu-7JLakEIl_L7nKHGTDHXfSywwqtwMo4L5zDlkFmxUYXcHWxJVPKN3A2C8Z3T2PK2-MmpVNyeNN92f-gHVW6RBZvKEAOk7tiZU92fP6R7yIiH1Votdy_kBFxVgSdIG5xpF30emN9f6V8s4M1n5fNy82qVIMzbzh70Xk5bwzJdArnN4rw&sockId=3" "error"="token expired"
```

Token Expires - 1660041672 --> Tue Aug 09 2022 10:41:12 GMT+0000
Log Line statement --> 2022/08/09 10:42:19

Going back to the original issue, it could be the case the epinio api pod crashed ... comes back up... the ui tries to reconnect and fails.

To fix this specific scenario the UI shouldn't try to reconnect with the original token... but fetch a new one first.

**Comment by jimmykarily** [comment]
@richard-cox with the PR merged, we can consider this fixed? Feel free to close this issue if that's the case, thanks.

**Comment by mmartin24** [comment]
Thanks @richard-cox  for the detailed steps. I tested this this issue on a k3d environment on version  `v1.1.0-16-g080ad496` using the same app originally posted (for some reason @richard-cox  I was a `Bad Gateway` error with yours; not sure if perhaps I am missing a env var )

Anyhow I could reproduce the error killing the `epinio-server` pod. 
![1](https://user-images.githubusercontent.com/37271841/185921032-ab1b25df-97c1-4544-bff4-b0f747a386fd.png)

As can be seen in the screenshot the same log as originally posted appears:

```
2022/08/22 11:24:55 EpinioServer: "msg"="Error" "error"="token expired"
```
However once down, the pod auto-recovers and the app is still accessible later:
![2](https://user-images.githubusercontent.com/37271841/185921095-7c559698-f292-4bc7-bad8-362a8a648372.png)

Since the app recovers from the `token expired` error, it seems to be ok.
Closing issue
